lib/Sema/CSApply.cpp

```diff
@@ -7543,6 +7543,16 @@ bool ConstraintSystem::applySolutionFix(Expr *expr,
                                                         DC);
     if (!useAs && !useAsBang)
       return false;
+    
+    // If we're performing pattern matching, "as" means something completely different...
+    if (auto binOpExpr = dyn_cast<BinaryExpr>(expr)) {
+      auto overloadedFn = dyn_cast<OverloadedDeclRefExpr>(binOpExpr->getFn());
+      if (overloadedFn && overloadedFn->getDecls().size() > 0) {
```

Comment:
Maybe size() == 1?

Reply:
This is consciously copying the code for matching "~=" that is in FailureDiagnosis::diagnoseTypeMemberOnInstanceLookup(). In practice the size is == 1, but the check is just for safety in getting any value decl. Should I change both places to be == 1, you think?

Reply:
@slavapestov Actually, looking at this again, we're hitting here before types are resolved on the expression in question, so the overload declarations include all known ~= implementations, not just one. So ```size() > 0``` is the correct check.

Reply:
People can call `~=` directly. Is there really not a better way to tell that we're in a `case` context?

cc @xedin